integration-tests/src/test_loop/tests/resharding_v3.rs

```diff
@@ -81,6 +82,115 @@ fn check_state_shard_uid_mapping_after_resharding(client: &Client, parent_shard_
     }
 }
 
+struct TestReshardingParameters {
+    chunk_ranges_to_drop: HashMap<ShardUId, std::ops::Range<i64>>,
+    accounts: Vec<AccountId>,
+    clients: Vec<AccountId>,
+    block_and_chunk_producers: Vec<AccountId>,
+    initial_balance: u128,
+    epoch_length: BlockHeightDelta,
+    /// Custom behavior executed at every iteration of test loop.
+    loop_action: Option<Box<dyn Fn(&mut TestLoopData, TestLoopDataHandle<ClientActorInner>)>>,
+}
+
+impl Default for TestReshardingParameters {
+    fn default() -> Self {
+        let initial_balance = 1_000_000 * ONE_NEAR;
+        let epoch_length = 6;
+        let accounts =
+            (0..8).map(|i| format!("account{}", i).parse().unwrap()).collect::<Vec<AccountId>>();
+        // #12195 prevents number of BPs bigger than `epoch_length`.
+        let clients = vec![accounts[0].clone(), accounts[3].clone(), accounts[6].clone()];
+        let block_and_chunk_producers = clients.clone();
+        let loop_action = None;
+        Self {
+            chunk_ranges_to_drop: HashMap::new(),
+            accounts,
+            clients,
+            block_and_chunk_producers,
+            initial_balance,
+            epoch_length,
+            loop_action,
+        }
+    }
+}
+
+impl TestReshardingParameters {
+    fn chunk_ranges_to_drop(
+        mut self,
+        chunk_ranges_to_drop: HashMap<ShardUId, std::ops::Range<i64>>,
+    ) -> Self {
+        self.chunk_ranges_to_drop = chunk_ranges_to_drop;
+        self
+    }
+
+    #[allow(unused)]
+    fn clients(mut self, clients: Vec<AccountId>) -> Self {
+        self.clients = clients;
+        self
+    }
+
+    #[allow(unused)]
+    fn block_and_chunk_producers(mut self, block_and_chunk_producers: Vec<AccountId>) -> Self {
+        self.block_and_chunk_producers = block_and_chunk_producers;
+        self
+    }
+
+    #[allow(unused)]
+    fn loop_action(
+        mut self,
+        loop_action: Option<Box<dyn Fn(&mut TestLoopData, TestLoopDataHandle<ClientActorInner>)>>,
+    ) -> Self {
+        self.loop_action = loop_action;
+        self
+    }
+}
+
+// Returns a callable function that, when invoked inside a test loop iteration, can force the creation of a chain fork.
+#[cfg(feature = "test_features")]
+fn fork_before_resharding_block(
+    double_signing: bool,
+) -> Box<dyn Fn(&mut TestLoopData, TestLoopDataHandle<ClientActorInner>)> {
+    let done = Arc::new(std::sync::atomic::AtomicBool::new(false));
```

Comment:
Using atomic was a my inaccuracy in other test, `std::cell::Cell` also works. We need only mutability, not the thread safety.

Reply:
Good point, I'll switch to `Cell`!